packages/client/components/EditorSuggestions/EditorSuggestions.tsx

```diff
@@ -25,7 +25,7 @@ const MentionMenu = styled('div')({
 
 interface Props {
   active: number
-  handleSelect: (idx: number) => (e: React.MouseEvent) => void
+  handleSelect: (item: DraftSuggestion) => void
```

Comment:
Hmm I'm a little confused:
1. How does changing to `BaseSuggestion` allow us to now case the `handleSelect` function of `SuggestMentionableUsers.tsx`?
2. Doesn't this make for _less_ type safety in context? I.e. if this is only dealing with `BaseSuggestion`s, that means this could call `handleSelect` with a `BaseSuggestion` that _doesn't_ represent a `MentionSuggestion` or `TagSuggestion`. So even if we're maintaining some state about whether the suggestion passed to the select handler should be a mention or a tag, there's no longer a guarantee that it's either.

I think if this was all more object-oriented then making these types the base class would make sense, but not with how this is currently written.

Reply:
1. gah i did a shit job of explaining.

Here's the current scenario:
- prop type: `handleSelect: (item: DraftSuggestion) => void`
- the above breaks down into: `handleSelect: (item: MentionSuggestion | TagSuggestion) => void`
-  passed in value: `(item: MentionSuggestion) => void`

The reason why typescript gets angry at this is because `EditorSuggestions` wants a function that can handle `MentionSuggestion | TagSuggestion`. `SuggestMentionableUsers` is only passing in a function that can handle `MentionSuggestions`. In other words, it's passing in something that can only fulfill half the required contract... and that's true! the function being passed in cannot handle a `TagSuggestion`. In _most_ code, this control flow would be inverted, where the thing being passed in can handle _more_ possibilities than the thing requiring it (e.g. "I can handle Mentions & Tags, but my child only needs me to handle Mentions").

So, we have 2 options: 

- we can either say "yes, it can in fact handle `TagSuggestion`, which is what the casting says (although from a pedantic standpoint that's not really true)
- or we can say "I'll accept any type of suggestion, I'm not going to discriminate based on the different types", which is what using the `BaseSuggestion` says

There's a word for this in type systems & I can't for the life of me remember what it is... but I hope this kinda helps! 

2. 100%! it does offer less type safety, The casting put in place says "I can handle Mentions or Tags", which isn't true, that function doesn't even know Tags are a thing! Using BaseSuggestion says  "I can handle _any_ Suggestion, but I'll treat them all the same". Less type safe, but more accurate. 

you're totally on point about OOP, we're writing functional code using a prototypical language with an OOP type system. it's a wild ride :sweat_smile:

Reply:
Okay, I think I understand a little better now. Still, in practice, we'd need to downcast from `BaseSuggestion` to `MentionSuggestion` in the handler. Instead of simply changing this to `BaseSuggestion`, I added generics (`T extends BaseSuggestion`) so passing in a bunch of `MentionSuggestion`s will mean the `handleSelect` gets called with a `MentionSuggestion`.